hudi-flink/src/main/java/org/apache/hudi/sink/partitioner/BucketAssignFunction.java

```diff
@@ -271,34 +272,39 @@ private void loadRecords(String partitionPath) throws Exception {
     final int parallelism = getRuntimeContext().getNumberOfParallelSubtasks();
     final int maxParallelism = getRuntimeContext().getMaxNumberOfParallelSubtasks();
     final int taskID = getRuntimeContext().getIndexOfThisSubtask();
+    final long currentTimeMillis = System.currentTimeMillis();
     for (HoodieBaseFile baseFile : latestBaseFiles) {
+      String commitTime = baseFile.getCommitTime();
+      long commitTimeLong = instantDateFormat.parse(commitTime).getTime();
+      // file was expired, needn't to fetch hoodieKeys
+      if (ttlTime > 0 && (commitTimeLong + ttlTime) < currentTimeMillis) {
+        continue;
+      }
       final List<HoodieKey> hoodieKeys;
       try {
         hoodieKeys =
             fileUtils.fetchRecordKeyPartitionPath(hadoopConf, new Path(baseFile.getPath()));
       } catch (Exception e) {
         // in case there was some empty parquet file when the pipeline
-        // crushes exceptionally.
-        LOG.error("Error when loading record keys from file: {}", baseFile);
-        continue;
+        // this exception is every important may cause duplication data , exception shouldn't be swallowed
+        LOG.error("Error when loading record keys from file: {}", baseFile, e);
+        throw e;
       }
-      hoodieKeys.forEach(hoodieKey -> {
+      hoodieKeys.stream().filter(hoodieKey -> recordKey.equals(hoodieKey.getRecordKey())).forEach(hoodieKey -> {
```

Comment:
can we take one hoodieKey only(if there), and avoid the foreach.

Reply:
since we are scanning all the files in the partition, can we stop this loop when we find the location of the recordKey